service/tcp.go

```diff
@@ -148,19 +148,43 @@ func (s *tcpService) SetTargetIPValidator(targetIPValidator onet.TargetIPValidat
 	s.targetIPValidator = targetIPValidator
 }
 
+// SOCKS address types as defined in RFC 1928 section 5.
+const (
+	AtypIPv4       = 1
+	AtypDomainName = 3
+	AtypIPv6       = 4
+)
+
 func dialTarget(tgtAddr socks.Addr, proxyMetrics *metrics.ProxyMetrics, targetIPValidator onet.TargetIPValidator) (onet.DuplexConn, *onet.ConnectionError) {
-	tgtTCPAddr, err := net.ResolveTCPAddr("tcp", tgtAddr.String())
-	if err != nil {
-		return nil, onet.NewConnectionError("ERR_RESOLVE_ADDRESS", fmt.Sprintf("Failed to resolve target address %v", tgtAddr.String()), err)
-	}
-	if err := targetIPValidator(tgtTCPAddr.IP); err != nil {
-		return nil, err
+	switch tgtAddr[0] {
+	case AtypDomainName:
+		tgtHost := string(tgtAddr[2 : 2 + int(tgtAddr[1])])
+		tgtIPs, err := net.LookupIP(tgtHost)
+		if err != nil {
+			return nil, onet.NewConnectionError("ERR_RESOLVE_ADDRESS", fmt.Sprintf("Failed to resolve target address %v", tgtAddr.String()), err)
+		}
+		for _, ip := range tgtIPs {
+			if err := targetIPValidator(ip); err != nil {
+				return nil, err
+			}
+		}
+	case AtypIPv4:
+		tgtIP := net.IP(tgtAddr[1 : 1 + net.IPv4len])
+		if err := targetIPValidator(tgtIP); err != nil {
+			return nil, err
+		}
+	case AtypIPv6:
+		tgtIP := net.IP(tgtAddr[1 : 1 + net.IPv6len])
+		if err := targetIPValidator(tgtIP); err != nil {
+			return nil, err
+		}
 	}
 
-	tgtTCPConn, err := net.DialTCP("tcp", nil, tgtTCPAddr)
+	tgtConn, err := net.Dial("tcp", tgtAddr.String())
```

Comment:
Unfortunately I don't think this is safe.  In a [DNS rebinding](https://en.wikipedia.org/wiki/DNS_rebinding) attack, the domain could initially resolve to a harmless destination, and then quickly be changed to resolve to a forbidden IP before being re-resolved by this call.

AFAIK, the only way to get proper dual-stack behavior is therefore to replicate the dual-stack/happy-eyeballs logic.

Reply:
Another solution would be to change the destination IP restriction to an opt-in setting. It's a lot easier to implement, maintain, and in my opinion, a lot better. By aligning our behavior with other popular Shadowsocks server implementations, there won't be any surprises to new users.

I often find myself in the need of accessing the private network of the server hosting an Outline server. Due to the restriction, I have to set up another non-Outline Shadowsocks server or resort to other VPN solutions. I'm pretty sure many other users also find it frustrating: #46.

Reply:
Opening up local host and network access is a major security flaw. The fact that other implementation are vulnerable by default means that they should change. Outline will stick to the safer setting.

We can consider having an opt-in way of enabling access to the local network. Note, however, that that's also a significant vulnerability when you are on cloud, since you can often access metadata services. On DO you would be able to retrieve the tags for the droplet and retrieve the secret for the management API.

This feels like an advanced need for advanced users that could complicate the product and increase risk. We will probably not add to the Outline Manager, but we could consider an allow list in the outline-ss-server config.